src/cryptonote_core/service_node_list.h

```diff
@@ -34,17 +34,25 @@
 #include "serialization/serialization.h"
 #include "cryptonote_basic/cryptonote_basic_impl.h"
 #include "cryptonote_core/service_node_rules.h"
+#include "cryptonote_core/uptime_proof.h"
 #include "cryptonote_core/service_node_voting.h"
 #include "cryptonote_core/service_node_quorum_cop.h"
 #include "common/util.h"
 
+#include <lokimq/bt_serialize.h>
+
 namespace cryptonote
 {
 class Blockchain;
 class BlockchainDB;
 struct checkpoint_t;
 }; // namespace cryptonote
 
+namespace uptime_proof
+{
+  class Proof;
+}
+
```

Comment:
This forward declaration doesn't seem necessary as you're including the `uptime_proof.h` header with the full declaration in it.

Reply:
So this is the only review note that I haven't been able to resolve. There is a circular reference between `service_node_list` and `uptime_proof` because signing the uptime proof uses `service_nodes::service_node_keys`. I could move the actual signing to another function within `service_node_lists` but feel the signing works better within the `uptime_proof` constructor. Thoughts?

Reply:
Can you keep the `Proof` forward declaration and instead remove the `#include "cryptonote_core/uptime_proof.h"` from the header (and put it into `service_node_list.cpp` instead)?  That should break the circular include, but I may be missing something else that s_n_l.h requires from `uptime_proof.h`.